samples/common/net_connect.c

```diff
@@ -12,10 +12,65 @@ LOG_MODULE_REGISTER(golioth_samples, LOG_LEVEL_DBG);
 #include <zephyr/net/net_if.h>
 #include <zephyr/net/dhcpv4.h>
 
+struct wait_data {
+	struct k_sem sem;
+	struct net_mgmt_event_callback cb;
+};
+
+static void event_cb_handler(struct net_mgmt_event_callback *cb,
+			     uint32_t mgmt_event,
+			     struct net_if *iface)
+{
+	struct wait_data *wait = CONTAINER_OF(cb, struct wait_data, cb);
+
+	if (mgmt_event == cb->event_mask) {
+		k_sem_give(&wait->sem);
+	}
+}
+
+static void wait_for_iface_up_event(struct net_if *iface)
+{
+	struct wait_data wait;
+
+	wait.cb.handler = event_cb_handler;
+	wait.cb.event_mask = NET_EVENT_IF_UP;
+
+	k_sem_init(&wait.sem, 0, 1);
+	net_mgmt_add_event_callback(&wait.cb);
+
+	if (net_if_is_up(iface)) {
+		goto del_event_cb;
+	}
+
+	k_sem_take(&wait.sem, K_FOREVER);
+
+del_event_cb:
+	net_mgmt_del_event_callback(&wait.cb);
+}
+
+static void wait_for_iface_up_poll(struct net_if *iface)
+{
+	while (!net_if_is_up(iface)) {
+		LOG_DBG("Interface is not up yet!");
+		k_sleep(K_MSEC(100));
+	}
+}
+
+static void wait_for_iface_up(struct net_if *iface)
+{
+	if (IS_ENABLED(CONFIG_NET_MGMT_EVENT)) {
+		wait_for_iface_up_event(iface);
+	} else {
+		wait_for_iface_up_poll(iface);
+	}
+}
+
 void net_connect(void)
 {
 	struct net_if *iface = net_if_get_default();
 
+	wait_for_iface_up(iface);
```

Comment:
Might be helpful to put a log message here, in case it gets stuck.

Reply:
Added `LOG_INF("Waiting for interface to be up");`. Is that fine?

Reply:
Perfect, thanks